Give builtin actions a platform

Silic0nS0ldier opened the pull request:
Continuing the work started in #22925.

This PR creates a special platform for builtin actions like `MiddlemanAction`, `SymlinkAction`, `SolibSymlinkAction`, and several others via `ActionOwner#SYSTEM_ACTION_OWNER` (e.g. `BazelWorkspaceStatusAction`, `CoverageReportAction`) in order to see the removal of the `@Nullable` constraint from `Action#getExecutionPlatform`.

The introduced builtins platform (`@_builtins//virtual:builtin_platform`) and the existing empty platform (`@_builtins//virtual:empty_platform`) have been given labels to aid debugging in the event they show up in an expected place.

These changes relate to the [Execution Platform Scoped Spawn Strategies](https://github.com/bazelbuild/proposals/blob/2c01184c39b29fba4a0bd2c0ea41c75ee0a6e444/designs/2023-06-04-exec-platform-scoped-spawn-strategies.md) proposal which seeks to introduce optional restrictions on how a action can be spawn on a given platform (e.g. to ensure spawns for Linux do not run on a macOS host).

cc @katre

katre reviewed (changes requested):
I'm checking this out on our internal tests but it looks good so far.

Silic0nS0ldier commented on src/main/java/com/google/devtools/build/lib/actions/ActionOwner.java:
```diff
@@ -39,14 +39,14 @@ public abstract class ActionOwner {
   public static final ActionOwner SYSTEM_ACTION_OWNER =
       createDummy(
-          /* label= */ null,
+          /* label= */ Label.parseCanonicalUnchecked("@_builtins//virtual:system_action_owner"),
```
Makes sense, although I've a minor concern that having a vestigial (unused) mirror of this platform could lead to behavioral differences if used at some later date. Since there would technically be 2 separate definitions (one in Java, another in Starlark).

Silic0nS0ldier commented:
@katre I've made some changes since the last review. Do you have any updates on this?

This is blocking #24265 which depends on all spawnable actions having a platform.

katre commented:
Sorry for the delay, let me restart testing this.

katre reviewed (approved):
Sorry for the long delay. This looks good, I'm going to handle merging myself. I need to run further tests internally, as well as make sure there aren't any places in internal code that need to be updated. How did you identify methods to change to return the empty platform?

Silic0nS0ldier commented:
> How did you identify methods to change to return the empty platform?

Searching for common patterns mainly. e.g.
- `executionPlatform=` (comment)
- `PlatformInfo getExecutionPlatform` (method)

The few outliers I either stumbled across or I noted down previously while prototyping https://github.com/bazelbuild/proposals/blob/80b77791301bd4778e712ad1c8e17e3aeb1b050a/designs/2023-06-04-exec-platform-scoped-spawn-strategies.md

`src/test/java/com/google/devtools/build/lib/runtime/ExecutionGraphModuleTest.java` however was a test failure. Beyond that I cannot recall.

katre commented:
Okay, importing this has been a ride. Thanks for your patience, thanks @comius for internal review (which is still ongoing).

Due to @comius's review, I've moved the platform definition to `@bazel_tools//tools:internal_platform` and added a constant to reference it (because it is something else entirely inside google).

There have also been a few other changes and cleanups.

I'm going to continue the internal review but won't actually commit anything. @Silic0nS0ldier, please take a look and comment on #25196 if you see any problems with anything I've changed.
